helix-gateway/src/main/java/org/apache/helix/gateway/util/GatewayCurrentStateCache.java

```diff
@@ -0,0 +1,144 @@
+
+
+public class GatewayCurrentStateCache {
+  String _clusterName;
+
+  // A cache of current state. It should be updated by the HelixGatewayServiceChannel
+  // instance -> instances assignments
+  Map<String, Map<String, String>> _currentStateMap;
+
+  // A cache of target state.
+  // instance -> assignments
+  Map<String, Map<String, String>> _targetStateMap;
+
+  boolean _targetStateChanged = false;
+
+  public GatewayCurrentStateCache(String clusterName) {
+    _clusterName = clusterName;
+    _currentStateMap = new ConcurrentHashMap<>();
+    _targetStateMap = new ConcurrentHashMap<>();
+  }
+
+  public String getCurrentState(String instance, String shard) {
+    return _currentStateMap.get(instance).get(shard);
+  }
+
+  /**
+   * Update the cached current state of instances in a cluster, and return the diff of the change.
+   * @param newCurrentStateMap The new current state map of instances in the cluster
+   * @return
+   */
+  public Map<String, Map<String, String>> updateCacheWithNewCurrentStateAndGetDiff(
+      Map<String, Map<String, String>> newCurrentStateMap) {
+    Map<String, Map<String, String>> diff = null;
+    for (Map.Entry<String, Map<String, String>> entry : newCurrentStateMap.entrySet()) {
+      String instance = entry.getKey();
+      Map<String, String> newCurrentState = entry.getValue();
+      Map<String, String> oldCurrentState = _currentStateMap.get(instance);
+      if (oldCurrentState == null || !oldCurrentState.equals(newCurrentState)) {
+        if (diff == null) {
+          diff = new HashMap<>();
+        }
+        if (oldCurrentState == null) {
+          diff.put(instance, newCurrentState);
+          continue;
+        }
+        for (String shard : newCurrentState.keySet()) {
+          if (oldCurrentState.get(shard) == null || !oldCurrentState.get(shard).equals(newCurrentState.get(shard))) {
+            diff.computeIfAbsent(instance, k -> new HashMap<>()).put(shard, newCurrentState.get(shard));
+          }
+        }
+      }
+    }
+    _currentStateMap = newCurrentStateMap;
+    return diff;
+  }
+
+  public void updateCacheWithCurrentStateDiff(Map<String, Map<String, String>> currentStateDiff) {
+    updateShardStateMapWithDiff(currentStateDiff, _currentStateMap);
+  }
+
+  /**
+   * Udate the target state with the changed target state maps.
+   * All existing target states remains the same
+   * @param targetStateChangeMap
+   */
+  public void updateTargetStateWithDiff(Map<String, Map<String, String>> targetStateChangeMap) {
+    _targetStateChanged = updateShardStateMapWithDiff(targetStateChangeMap, _targetStateMap);
+  }
+
+  public boolean isTargetStateChanged() {
+    return _targetStateChanged;
+  }
+
+  public void resetTargetStateChanged() {
+    _targetStateChanged = false;
+  }
+
+  public Map<String, Map<String, String>> getTargetStateMap() {
+    return _targetStateMap;
+  }
+
+  public String serializeTargetAssignments() {
+    ObjectMapper mapper = new ObjectMapper();
```

Comment:
Can we have a java class as the object and using annotation for parsing? That can help the standardize the format.